### PR TITLE
feat(minimessage): validate markup and placeholders

### PR DESCRIPTION
**Pull request by LMLiam**
## Summary

Adds a structured validation layer for MiniMessage markup and typed placeholders. Introduces a `ValidationResult`/`MiniMessageDiagnostic` model, a two-pass algorithm (strict-mode parse + recording resolver), and a `MiniTemplate.validate()` convenience entry point.

Key behaviours:
- **Malformed/unclosed tag detection** — identifies unrecognised or unclosed tags with source position information (`MiniMessageDiagnostic.Kind.MALFORMED_TAG`). Note: in strict mode, unclosed standard tags (e.g. `<bold>` without `</bold>`) are also reported as malformed — this is the intentional strict-mode behaviour of the underlying MiniMessage parser.
- **Missing/extra placeholder detection** — cross-checks declared template placeholders against those referenced in the markup; reports `MISSING_PLACEHOLDER` and `EXTRA_PLACEHOLDER` diagnostics.
- **Standard-tag filtering** — known MiniMessage tags (colours, decorations, etc.) are excluded from the unknown-tag check so only genuinely unknown references are flagged.
- **`MiniTemplate.validate()`** — convenience method; consumers call validate on a template instance without accessing the parser internals.
- **Basis for #49 / #51** — the diagnostic model is designed to power IDE inspections and error reporting in future issues, but neither is implemented here.

CHANGELOG is release-please-owned and is not hand-edited.

## Related Issues

Closes #28

## DSL Impact

No breaking changes to the existing `MiniTemplate` DSL surface. Adds `validate(): ValidationResult` and the supporting `ValidationResult` / `MiniMessageDiagnostic` types to the `minimessage` module.

## Rationale

Plugin authors need early feedback when a message template contains a typo'd tag or references a placeholder that was never declared. Without a validation pass, errors only surface at send-time (or not at all). This gives authors a compile-adjacent check they can run in tests or at startup.

## Testing

20 Kotest test cases covering:
- Valid templates with and without placeholders
- Unclosed standard tags (strict-mode malformed behaviour)
- Unrecognised/misspelled tags
- Missing placeholders (declared but not used in markup)
- Extra placeholders (used in markup but not declared)
- Mixed diagnostic combinations
- `MiniTemplate.validate()` convenience entry point

Run with `./gradlew test`. README docs added under the MiniMessage validation section.

## Checklist

- [x] Linked related issue
- [x] Updated README and relevant `/docs` pages
- [x] Verified examples build and run

### CODE REVIEW

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageDiagnostic.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package io.github.lmliam.kotventure.minimessage.validation
+
+import net.kyori.adventure.text.minimessage.ParsingException
+
+/**
+ * A single diagnostic produced by [io.github.lmliam.kotventure.minimessage.validate].
+ *
+ * Sealed so callers can exhaustively handle every diagnostic kind.
+ */
+public sealed class MiniMessageDiagnostic {
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageDiagnostic.kt]: Shouldn't this just be a sealed interface, rather than a class?

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package io.github.lmliam.kotventure.minimessage.validation
+
+/**
+ * The result of a [io.github.lmliam.kotventure.minimessage.validate] call.
+ *
+ * Either [Success] (markup is well-formed and every spec placeholder is present) or
+ * [Failure] (one or more [MiniMessageDiagnostic] were found).
+ */
+public sealed class ValidationResult {
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt]: Shouldn't this also be an interface?

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package io.github.lmliam.kotventure.minimessage.validation
+
+/**
+ * The result of a [io.github.lmliam.kotventure.minimessage.validate] call.
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt]: Can we avoid full references like this?

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * `true` when this result is [ValidationResult.Success]; `false` otherwise.
+ */
+public val ValidationResult.isSuccess: Boolean
+    get() = this is ValidationResult.Success
+
+/**
+ * `true` when this result is [ValidationResult.Failure]; `false` otherwise.
+ */
+public val ValidationResult.isFailure: Boolean
+    get() = this is ValidationResult.Failure
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt]: Why do these need to be extension methods?

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt [hunk 5]
```diff
@@ -0,0 +1,165 @@
+package io.github.lmliam.kotventure.minimessage
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt:1]: This class seems to fail SRP

**Review by LMLiam — COMMENTED** [review]
## Self-review (Round 1) — automated + manual diff walk

**CI status:** All checks pass. Build, Test, and Lint ✅ | Qodana ✅ | GitGuardian ✅ | Validate PR Title/Commits ✅ | CodeRabbit ✅ | Qodana for JVM (skipping — known neutral noise per repo memory).

**Local verification:** `./gradlew :minimessage:test` → 20/20 pass, 0 failures. `./gradlew build` → SUCCESS. `./gradlew ktlintCheck spotlessCheck` → SUCCESS.

### Findings

Three findings below — two are genuine defects, one is a missing test. CodeRabbit already flagged the two defects independently; confirming them as real. Inline comments added per file+line.

#### 🔴 F1 — Pass 2 recording parse has no catch: breaks no-exception contract (Important)

`MiniMessageValidation.kt:108` — `MiniMessage.miniMessage().deserialize(markup, combined)` in `detectPlaceholderMismatches` is unguarded. Adventure's lenient parser is designed not to throw `ParsingException`, but documented cases exist where malformed tag arguments (e.g. unclosed double quotes in tag attributes) cause `StringIndexOutOfBoundsException` from within the parser internals (PaperMC/adventure#1011). The `validate()` KDoc promises no exception escapes the caller; Pass 2 can violate this.

Suggested fix: wrap the Pass 2 `deserialize` call in a `try/catch(RuntimeException)`, and on catch either propagate the diagnostic as a `MalformedTag` or return what was collected so far. A simpler safe option is `try { … } catch (_: RuntimeException) { /* lenient parser bug; ignore and proceed with partial results */ }`.

#### 🟡 F2 — Standard-tag name collision: spec placeholder named 'gold'/'bold'/etc. produces false MissingPlaceholder (Important)

`MiniMessageValidation.kt:158` — The recording resolver filters out any tag name for which `TagResolver.standard().has(name)` is true. If the spec contains `placeholder<String>("gold")` (technically valid per the `[!?#]?[a-z0-9_-]+` regex), Pass 2 never records "gold" in `encounteredNames`, so `<gold>` in markup is falsely reported as `MissingPlaceholder("gold")` even though the tag IS present.

Severity: the regex does not block standard tag names, so callers can accidentally hit this. Suggested fix: either (a) widen the filter to record standard-tag names that are also in the spec, or (b) reject placeholder names that collide with standard tags at declaration time (in `MiniMessagePlaceholder` or the `placeholder()` factory). Option (b) is safer and consistent with early-rejection design.

#### 🟡 F3 — Missing test for LOCATION_UNKNOWN path (Nice-to-have, from task plan)

The task plan explicitly required a test `"MalformedTag startIndex is LOCATION_UNKNOWN when position unavailable"`. 19 of the 17 planned tests are present; this one is absent. The code correctly passes `e.startIndex()`/`e.endIndex()` through unconditionally, and these return `LOCATION_UNKNOWN` when Adventure doesn't know the position — but there is no test that exercises that path. Acknowledging this may be hard to trigger without knowing a specific Adventure input that causes LOCATION_UNKNOWN; if no such input is readily available, a note explaining why the test was dropped would be useful.

### Per-AC verdict

| AC | Status | Evidence |
| -- | ------ | -------- |
| AC1 malformed/unclosed tags detected with location | ✅ Satisfied | Pass 1 strict parse; MalformedTag with startIndex/endIndex; LOCATION_UNKNOWN mirrored from Adventure |
| AC2 missing placeholders reported (spec→markup) | ✅ Satisfied | Pass 2 recording resolver; specNames - tagsInMarkup |
| AC3 extra placeholders reported (markup→spec) | ⚠️ Partially — F2 above | tagsInMarkup - specNames, but standard-tag-name placeholder collision causes false negative |
| AC4 structured result, no exception to caller | ⚠️ Partially — F1 above | Pass 1 is guarded; Pass 2 can throw RuntimeException for edge-case malformed markup |
| AC5 tests cover positive and negative paths | ✅ Mostly satisfied | 20 tests, all pass; LOCATION_UNKNOWN path test absent (F3) |
| AC6 docs updated (README); CHANGELOG not edited | ✅ Satisfied | README Validation section added; no CHANGELOG edits |

### What's good

- Algorithm design (two-pass, self-closing spec resolver to prevent false strict-mode positives) is solid and well-explained.
- KDoc on all public API; no wildcard imports; no ; no Java-isms; explicit visibility everywhere.
- Test naming is descriptive; assertions are specific; all 20 pass locally.
-  visibility change is minimal (private → internal, justified).
- Diagnostic ordering guarantee is documented in both KDoc and README.
- README examples use properly closed standard tags throughout (strict-mode-correct).
- No scope creep — no #49/#51 integration, no unrelated refactors.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+ * - extra: recorded names absent from the spec (in markup-encounter order).
+ */
+private fun detectPlaceholderMismatches(
+    markup: String,
+    spec: List<MiniMessagePlaceholder<*>>,
+): Pair<List<MiniMessageDiagnostic.MissingPlaceholder>, List<MiniMessageDiagnostic.ExtraPlaceholder>> {
+    val recorder = RecordingTagResolver()
+    val combined = TagResolver.resolver(TagResolver.standard(), recorder)
+    MiniMessage.miniMessage().deserialize(markup, combined)
+
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: **F1 — Pass 2 unguarded deserialize can break the no-exception contract**

`MiniMessage.miniMessage().deserialize(markup, combined)` on this line has no `catch` block. The lenient parser is designed not to throw `ParsingException`, but Adventure's parser internals can throw `StringIndexOutOfBoundsException` for edge-case malformed inputs (e.g. unclosed double-quote in a tag argument — PaperMC/adventure#1011). The `validate()` KDoc promises no exception escapes the caller; this line can violate that promise.

Suggested minimal fix:
```kotlin
try {
    MiniMessage.miniMessage().deserialize(markup, combined)
} catch (_: RuntimeException) {
    // Adventure parser bug on edge-case malformed input — proceed with partial results.
}
```

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    /** Tag names encountered in the markup, in the order they were first seen. */
+    val encounteredNames: LinkedHashSet<String> = LinkedHashSet()
+
+    override fun resolve(
+        name: String,
+        arguments: ArgumentQueue,
+        ctx: Context,
+    ): Tag? {
+        if (!TagResolver.standard().has(name)) {
+            encounteredNames.add(name)
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: **F2 — Standard-tag-name placeholder causes false MissingPlaceholder**

The filter `!TagResolver.standard().has(name)` correctly prevents standard tags like `<gold>` from appearing as extra placeholders. However, it also silently drops any placeholder whose name coincidentally matches a standard tag. Because `MiniMessagePlaceholder`'s name regex (`[!?#]?[a-z0-9_-]+`) does not block names like `gold`, `bold`, `italic`, or `rainbow`, the following silently gives a false `MissingPlaceholder("gold")`:

```kotlin
val gold = placeholder<String>("gold")
validate("<gold>text", listOf(gold))
// → Failure([MissingPlaceholder("gold")]) — wrong; <gold> IS in the markup
```

Decide whether to (a) record a name if it is in the spec even when standard also knows it, or (b) reject spec names that collide with standard tags at placeholder-declaration time. Option (b) aligns with the early-rejection design already present for duplicate names.

File: modules/minimessage/src/test/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidationTest.kt [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+                    spec = listOf(player),
+                )
+
+                val failure = result.shouldBeInstanceOf<ValidationResult.Failure>()
+                val malformed = failure.diagnostics.filterIsInstance<MiniMessageDiagnostic.MalformedTag>()
+                malformed shouldHaveSize 1
+                malformed[0].message.isNotEmpty() shouldBe true
+            }
+
+            "MalformedTag.LOCATION_UNKNOWN sentinel is sourced from Adventure not hardcoded" {
```
Comment (LMLiam) [inline modules/minimessage/src/test/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidationTest.kt:109]: **F3 — Missing test for LOCATION_UNKNOWN path (planned but absent)**

The task plan included a test named `"MalformedTag startIndex is LOCATION_UNKNOWN when position unavailable"`. 19 of the 17 planned tests are present; this one is not. The code always passes `e.startIndex()`/`e.endIndex()` through to the diagnostic (the path works correctly if Adventure returns the sentinel), but there is no test that proves it.

If triggering LOCATION_UNKNOWN requires a specific Adventure input that's hard to find, a comment here (or in the test plan) explaining that would close the gap without requiring the test itself.

File: README.md [hunk 9]
```diff
@@ -312,6 +312,99 @@ val forSam = WelcomeTemplate {
...
+import io.github.lmliam.kotventure.minimessage.validation.MiniMessageDiagnostic
+import io.github.lmliam.kotventure.minimessage.validation.ValidationResult
+import net.kyori.adventure.text.Component
+
+val player = placeholder<Component>("player")
+val count = placeholder<Int>("count")
+
+// Success — well-formed markup, all placeholders present, standard tags closed
+val result = validate(
+    markup = "<gold>Welcome <player></gold>, you have <count> messages",
```
Comment (LMLiam) [inline README.md]: I'm not sure whether `markup` is the right term

File: README.md [hunk 10]
```diff
@@ -312,6 +312,99 @@ val forSam = WelcomeTemplate {
...
+    spec = listOf(player, count),
```
Comment (LMLiam) [inline README.md]: Same here, I don't know if `spec` is the right term

File: README.md [hunk 11]
```diff
@@ -312,6 +312,99 @@ val forSam = WelcomeTemplate {
...
+    markup = "<gold>Hello <player></gold> <mystery>",
+    spec = listOf(player),
+)
+// extraResult == ValidationResult.Failure(
+//   diagnostics = [ExtraPlaceholder("mystery")]
+// )
+
+// Inspect diagnostics with an exhaustive when
+when (val r = missingResult) {
+    is ValidationResult.Success -> println("All good!")
```
Comment (LMLiam) [inline README.md]: Can we avoid having our prime examples using `println`?

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt (hunk 6 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: Fixed in ebb4697. Wrapped the lenient `deserialize` call in `detectPlaceholderMismatches` in `try { … } catch (_: RuntimeException) { }` so any Adventure-internal parser bug (e.g. `StringIndexOutOfBoundsException` — PaperMC/adventure#1011) is swallowed and the recording side-effects gathered so far are still used. Applied the same guard consistently to `detectMalformedTags` for non-`ParsingException` `RuntimeException`s, preserving the no-throw contract of `validate()` in both passes. Regression test added: _"validate returns a result rather than throwing for markup that is well-formed in lenient mode"_.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt (hunk 7 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: Fixed in ebb4697. `RecordingTagResolver` now accepts the set of spec names and records a tag name when `name in specNames || !TagResolver.standard().has(name)`. This means a placeholder whose name collides with a standard tag (e.g. `gold`) is correctly recorded when the tag appears in the markup, preventing the false `MissingPlaceholder`. Standard tags that are NOT spec names continue to be excluded from the recorded set, so they never appear as extra placeholders. Regression tests added: _"placeholder named same as standard tag is not falsely reported as MissingPlaceholder when used in markup"_, _"placeholder named same as standard tag is reported as MissingPlaceholder when absent from markup"_, and _"standard tag not in spec is never reported as ExtraPlaceholder when spec has standard-named placeholder"_.

File: modules/minimessage/src/test/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidationTest.kt (hunk 8 above)
Comment (LMLiam) [reply modules/minimessage/src/test/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidationTest.kt:109]: Fixed in ebb4697. Two additions: (1) a new test _"MalformedTag start and end indices are passed through from Adventure without modification"_ asserts that `startIndex`/`endIndex` are forwarded verbatim from Adventure when it reports a location; (2) a comment near the test notes that Adventure 5.1.1 does not expose a public input that yields the `LOCATION_UNKNOWN` sentinel from the parser, so the LOCATION_UNKNOWN path cannot be tested with a concrete input — but the compile-time binding `LOCATION_UNKNOWN shouldBe ParsingException.LOCATION_UNKNOWN` (existing test) proves the constant is Adventure-derived rather than hardcoded.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt (hunk 6 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: You have mentioned the issue #1011 but this has been closed for a long time now.

Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: Good catch — fixed in 2921d97. Removed the `PaperMC/adventure#1011` reference from both the production comment and the test comment; replaced it with a plain description of the defensive rationale (the `RuntimeException` guard upholds the no-throw contract of `validate()` regardless of future parser changes).

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt (hunk 2 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt]: Converted to  in 7c9cdfa.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageDiagnostic.kt (hunk 1 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageDiagnostic.kt]: Converted to sealed interface in 7c9cdfa.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt (hunk 4 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt]: Moved to member properties with default getters on the sealed interface in 7c9cdfa. The extension-property declarations are removed.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt (hunk 3 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt]: Shortened to [validate] in 7c9cdfa. Same-package resolution makes the unqualified form valid.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt (hunk 5 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt:1]: Done in 7c9cdfa: RecordingTagResolver moved to validation/RecordingTagResolver.kt (internal), the two-pass algorithm moved to validation/MiniMessageValidator.kt (internal fun runValidation). MiniMessageValidation.kt now holds only the public entry points validate(input, placeholders) and MiniTemplate.validate().

File: README.md (hunk 9 above)
Comment (LMLiam) [reply README.md]: Renamed to input in 7c9cdfa (matches the existing mini(input) entry point). README and test named args updated in 3384b63.

File: README.md (hunk 10 above)
Comment (LMLiam) [reply README.md]: Renamed to placeholders in 7c9cdfa (matches MiniTemplate.declaredPlaceholders). README and test named args updated in 3384b63.

File: README.md (hunk 11 above)
Comment (LMLiam) [reply README.md]: Replaced println usage in 3384b63. The exhaustive-when example now maps diagnostics to strings and assigns the result, showing the pattern without side effects.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+     *   diagnostics appear first, then missing-placeholder diagnostics in spec declaration order,
+     *   then extra-placeholder diagnostics in the order the tags were encountered in the markup.
+     */
+    public data class Failure(
+        public val diagnostics: List<MiniMessageDiagnostic>,
+    ) : ValidationResult() {
+        init {
+            require(diagnostics.isNotEmpty()) { "Failure must carry at least one diagnostic." }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/ValidationResult.kt]: Implemented in 7c9cdfa: Failure now has a private primary constructor and a public companion invoke that calls .toList() to snapshot the list before storing it. Equality-based tests still pass since data class equality compares the list contents.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+ * - missing: spec names absent from the recorded tags (in spec declaration order).
+ * - extra: recorded names absent from the spec (in markup-encounter order).
+ */
+private fun detectPlaceholderMismatches(
+    markup: String,
+    spec: List<MiniMessagePlaceholder<*>>,
+): Pair<List<MiniMessageDiagnostic.MissingPlaceholder>, List<MiniMessageDiagnostic.ExtraPlaceholder>> {
+    val recorder = RecordingTagResolver()
+    val combined = TagResolver.resolver(TagResolver.standard(), recorder)
+    MiniMessage.miniMessage().deserialize(markup, combined)
```
[the remark this answers is not in the document]
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: Already addressed in ebb4697 (lenient parse wrapped in catch(RuntimeException)). The algorithm now also lives in validation/MiniMessageValidator.kt (7c9cdfa).

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt]: Already addressed in ebb4697: the RecordingTagResolver now records a name when it is in specNames OR not a standard tag, so a placeholder named 'gold' is recorded when <gold> appears in the markup and does not produce a false MissingPlaceholder. This is covered by regression tests in MiniMessageValidationTest. Algorithm now lives in validation/MiniMessageValidator.kt (7c9cdfa).

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidation.kt:1]: ktlintCheck and spotlessCheck (the project's enforced formatters per CONTRIBUTING/CLAUDE.md) pass clean on these files. The Qodana/IntelliJ-formatter alert is a known divergence and not actionable under the project's style config.

File: modules/minimessage/src/test/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidationTest.kt [hunk 14]
```diff
@@ -0,0 +1,338 @@
+package io.github.lmliam.kotventure.minimessage
```
[the remark this answers is not in the document]
Comment (LMLiam) [reply modules/minimessage/src/test/kotlin/io/github/lmliam/kotventure/minimessage/MiniMessageValidationTest.kt:1]: ktlintCheck and spotlessCheck (the project's enforced formatters per CONTRIBUTING/CLAUDE.md) pass clean on these files. The Qodana/IntelliJ-formatter alert is a known divergence and not actionable under the project's style config.

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageValidator.kt [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+private fun detectPlaceholderMismatches(
+    input: String,
+    placeholders: List<MiniMessagePlaceholder<*>>,
+): Pair<List<MiniMessageDiagnostic.MissingPlaceholder>, List<MiniMessageDiagnostic.ExtraPlaceholder>> {
+    val specNames = placeholders.map { it.name }.toSet()
+    val recorder = RecordingTagResolver(specNames)
+    val combined = TagResolver.resolver(TagResolver.standard(), recorder)
+    // The lenient parser is designed not to throw ParsingException, but Adventure's parser
+    // internals can still throw RuntimeException (e.g. StringIndexOutOfBoundsException) on
+    // edge-case malformed inputs. Guard defensively so validate() never propagates an exception;
```
Comment (LMLiam) [inline modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageValidator.kt]: Hasn't this been resolved in adventure?

File: README.md [hunk 16]
```diff
@@ -312,6 +312,97 @@ val forSam = WelcomeTemplate {
 //   → IllegalArgumentException: Template is missing required placeholder(s): [count].
 ```
 
+`validate(markup, spec)` checks a MiniMessage string against a declared placeholder spec and returns a structured
```
Comment (LMLiam) [inline README.md]: this still says markup and spec

File: modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageValidator.kt (hunk 15 above)
Comment (LMLiam) [reply modules/minimessage/src/main/kotlin/io/github/lmliam/kotventure/minimessage/validation/MiniMessageValidator.kt]: Good catch — yes, the specific crash that originally motivated this (adventure#1011, a StringIndexOutOfBoundsException) was fixed in Adventure 4.16.0, well before the 5.1.1 we use, so that exact reference was stale and I've removed it.

I've kept a guard here though, because `validate` documents a no-throw contract ("no exception escapes to the caller", AC4) and the lenient parser isn't contractually guaranteed exception-free for every input/version. To make it correct rather than merely safe, the strict pass now emits a `MalformedTag` (with `LOCATION_UNKNOWN`) instead of silently swallowing to an empty list — previously a crash-inducing input could have slipped through as `Success`. Done in 46eff40 / aad4f2b.

File: README.md (hunk 16 above)
Comment (LMLiam) [reply README.md]: Fixed in 46eff40 — the prose in this section now uses `validate(input, placeholders)` and refers to "input"/"placeholders" throughout, matching the renamed parameters. The `MiniTemplate.validate()` paragraph below no longer says "spec" either.
